### DISCUSSION TITLE
pause  consuming with auto ack true / rate limit with  with auto ack true

### DISCUSSION THREAD

**Discussion by parsibox** [rabbitmq/amqp091-go#247] [Q&A] [answered]
### Is your feature request related to a problem? Please describe.

i need to pause  consuming  in auto ack true mode
i do not want to lose consumed message that were in ram
i just want not get new message from my queue

### Describe the solution you'd like

i need to set a message rate for consume data from my queue in auto ack mode 
for example i add 10000 record to my queue and i want only get max 500 message per second 
i use prefetch 500 but it only work with auto ack false
please advice me

**Comment by lukebakken** [comment, marked as answer]
Thanks for using RabbitMQ and this AMQP client library.

> i need to set a message rate for consume data from my queue in auto ack mode

This is not supported. By design, auto ack mode does not limit the message rate to your consumer.

If you wish to limit the rate of messages to your consumer, and not lose messages that are in RAM, you **MUST** use manual ack mode.

If you think about it, this applies to ANY network protocol, not just AMQP.

**Comment by parsibox** [comment]
so at least is there anyway to stop consuming for auto ack mode without lose data?
if we use chanel.close is it safe?

**Reply by lukebakken** [reply]
> so at least is there anyway to stop consuming for auto ack mode without lose data?
> if we use chanel.close is it safe?

If your program crashes before processing all of the in-flight messages, you will lose them. RabbitMQ considers them processed immediately due to auto ack.

Again, this is due to how computers and networks work, and is not specific to RabbitMQ.

**Comment by parsibox** [comment]
no i manually want to stop consuming and after a while again start consuming 
but i do not want to lose any data
i want to count received data and for example if reached 1000 i stop consuming and actually i should receive more than 1000 data because some of them is in ram , i want processed that i items  after   stop consuming  and when all received data  processed  again start consuming

also if we have another method like Get that could received more than 1 data from queue it is very good for us
for example we only want to get maximum 1000 item in one call and again repeat this work ==> is there any method for it?

**Reply by lukebakken** [reply]
* Do _NOT_ use auto ack
* Set a prefetch of 1000
* Consume from your queue, and save the messages in RAM (a list, hash table, whatever)
* Process the 1000 messages.
* When done, acknowledge the entire batch by setting `multiple` to true and using the last delivery tag.
* Repeat

**Comment by parsibox** [comment]
thanks a lot
